drivers/ws281x/systick_gpio_ll.c

```diff
@@ -0,0 +1,134 @@
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* (+ NS_PER_SEC - 1): Rounding up, as T1H is the time that needs to distinctly
+ * longer than T0H.
+ *
+ * Then adding +1 extra, because the spin loop adds another layer of jitter. A
+ * more correct version would be to add the spin loop time before rounding (and
+ * then rounding up), but as that time is not available, spending one more
+ * cycle is the next best thing to do. */
+const int ticks_one = ((uint64_t)WS281X_T_DATA_ONE_NS * WS281X_TIMER_FREQ + NS_PER_SEC - 1)
+                    / NS_PER_SEC + 1;
+/* Rounding down, zeros are better shorter */
+const int ticks_zero = (uint64_t)WS281X_T_DATA_ZERO_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+/* No particular known requirements, but we're taking longer than that anyway
+ * because we don't clock the times between bits. */
+const int ticks_data = (uint64_t)WS281X_T_DATA_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+
+static void _systick_start(uint32_t ticks)
+{
+    /* disable SysTick, clear value */
+    SysTick->CTRL = 0;
+    SysTick->VAL = 0;
+    /* prepare value in re-load register */
+    SysTick->LOAD = ticks;
+    /* start and wait for the load value to be applied */
+    SysTick->CTRL = SysTick_CTRL_CLKSOURCE_Msk | SysTick_CTRL_ENABLE_Msk;
+    while (SysTick->VAL == 0) { /* Wait for SysTick to start and spin */ }
+}
+
+static void _systick_wait(void)
+{
+    while (!(SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk)) { /* busy wait */ }
+}
+
+void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
+{
+    assert(dev);
+
+    /* the high time for one can be as high as 5 seconds in practise, so
+     * rather be on the high side by adding a few CPU cycles. */
+    const uint32_t ticks_one = DIV_ROUND_UP((uint64_t)WS281X_T_DATA_ONE_NS * (uint64_t)coreclk(), NS_PER_SEC) + 16;
+    /* the low time should rather be on the short side, so rounding down */
+    const uint32_t ticks_zero = (uint64_t)(WS281X_T_DATA_ZERO_NS - 50) * (uint64_t)coreclk() / NS_PER_SEC;
+    /* the remaining time doesn't matter to much, should only be enough for the
+     * LEDs to detect the low phase. And not way to much to be detected as
+     * reset */
+    const uint32_t ticks_bit = DIV_ROUND((uint64_t)WS281X_T_DATA_NS * (uint64_t)coreclk(), NS_PER_SEC);
+
+    const uint8_t *pos = buf;
+    const uint8_t *end = pos + size;
+
+    gpio_port_t port = gpio_get_port(dev->params.pin);
+    uword_t mask = 1U << gpio_get_pin_num(dev->params.pin);
+
+    unsigned irq_state = irq_disable();
+    while (pos < end) {
+        uint8_t data = *pos;
+        for (uint8_t cnt = 8; cnt > 0; cnt--) {
+            uint32_t ticks_high = (data & 0x80) ? ticks_one : ticks_zero;
+            uint32_t ticks_low = ticks_bit - ticks_high;
+            _systick_start(ticks_high);
+            gpio_ll_set(port, mask);
+            _systick_wait();
+            gpio_ll_clear(port, mask);
+            _systick_start(ticks_low);
+            _systick_wait();
```

Comment:
ahh i see where something precise is need (high time) there are only: set; wait and clr. and for the low time the inprecise start; wait; and the next start is placed 

instead of reconfiguring the systick each time; i would start the systick once configured to some interval that cover high and low and wait for the count flag to raise to raise n times.

eg.:
```
set the systick to 100ns
for  each bit b  int byte do

set gpio 

if b == 1 
3 times  syst_wait()
else 
6 times syst_wait()
endif 

clear gpio 

if b == 1 
6 times  syst_wait()
else 
3 times syst_wait()
endif

done
```

this probably makes the whole thing less jittery

Reply:
That adds a branch in the timing sensitive phase and waiting for 300 ns on a STM32F072 is already pretty challenging with one time starting the systick timer. I doubt that you can start the systick timer three times within that time budget.

Reply:
not starting you just have it running

Reply:
the manuals that i found that had tolerances with the highs an lows would also be fine with the 300 ns and the 1: 2/3 1/3 and 0: 2/3 1/3 ( going close to the tolerances)

Reply:
What was the argument to split a single wait into three?

In any case: Each of the three waits would have to spin until a flag is set, clear the flag, and spin again. On the `nucleo-f072rb` the total wait is 13 cycles. Splitting that into three waits means 4 CPU cycles per wait. That is not possible to implement.

Reply:
Yeah, but a conditional branch will take more than one CPU cycle.

And again: Why divide a single wait? It adds lines or code, bytes to `.text`. And why into three waits? Wouldn't 42 or 1337 be much better numbers?

Reply:
because the has timing that are 1/4:3/4  or 1/3:2/3 

you can put the branch outside of the time critical path

when i wrote the first suggestion i wasn't aware that we run the nucleo 72 @ 12MHz (I am still not sure but 13 cycles in about a microseconds would fit 12MHz) 

but the reconfiguration of the systick is at least 7 instructions (if the register adresses are loaded) and a wait of unknown length

Reply:
it about not reconfiguring the systimer 

this spreads the waits a little better  so that they each have least amount of empty loops
```
    _systick_start( 4 );
    for (uint8_t cnt = 8; cnt > 0; cnt--) {
        int bit = !!(data & 0x80);
        if(!bit)
        {
            _systick_wait();
            gpio_ll_set();
            _systick_wait();
            gpio_ll_clear();
            _systick_wait();
        }else{
            _systick_wait();
            gpio_ll_set();
            _systick_wait();
            _systick_wait();
            gpio_ll_clear();
        }
        data <<= 1;
        _systick_wait();
    }
```